src/types.rs

```diff
@@ -136,3 +136,86 @@ impl LintPass for UnitCmp {
         }
     }
 }
+
+pub struct CastPass;
+
+declare_lint!(pub CAST_PRECISION_LOSS, Allow,
+              "casts that cause loss of precision, e.g `x as f32` where `x: u64`");
+declare_lint!(pub CAST_SIGN_LOSS, Allow,
+              "casts from signed types to unsigned types, e.g `x as u32` where `x: i32`");
+declare_lint!(pub CAST_POSSIBLE_TRUNCATION, Allow,
+              "casts that may cause truncation of the value, e.g `x as u8` where `x: u32`, or `x as i32` where `x: f32`");
+
+impl LintPass for CastPass {
+    fn get_lints(&self) -> LintArray {
+        lint_array!(CAST_PRECISION_LOSS,
+                    CAST_SIGN_LOSS,
+                    CAST_POSSIBLE_TRUNCATION)
+    }
+
+    fn check_expr(&mut self, cx: &Context, expr: &Expr) {
+        if let ExprCast(ref ex, _) = expr.node {
+            let (cast_from, cast_to) = (cx.tcx.expr_ty(&*ex), cx.tcx.expr_ty(expr));
+            if cast_from.is_numeric() && cast_to.is_numeric() && !in_external_macro(cx, expr.span) {
+                match (cast_from.is_integral(), cast_to.is_integral()) {
+                    (true, false) => {
+                        let from_nbits = match &cast_from.sty {
+                            &ty::TyInt(i) =>  4 << (i as usize),
+                            &ty::TyUint(u) => 4 << (u as usize),
+                            _ => 0
+                        };
+                        let to_nbits : usize = match &cast_to.sty {
+                            &ty::TyFloat(ast::TyF32) => 32,
+                            &ty::TyFloat(ast::TyF64) => 64,
+                            _ => 0
+                        };
+                        if from_nbits != 4 {
+                            // Handle TyIs/TyUs separately (pointer size is arch dependant)
+                            if from_nbits >= to_nbits {
+                                span_lint(cx, CAST_PRECISION_LOSS, expr.span,
+                                          &format!("converting from {0} to {1}, which causes a loss of precision \
+                                          			({0} is {2} bits wide, but {1}'s mantissa is only {3} bits wide)",
+                                                   cast_from, cast_to, from_nbits, if to_nbits == 64 {52} else {23} ));
+                            }
+                        }
+                    },
+                    (false, true) => {
+                        span_lint(cx, CAST_POSSIBLE_TRUNCATION, expr.span,
```

Comment:
Good question: how exactly is the floating point value "truncated" when cast to integral types? Apparently `1e100f64 as u64` is just zero...